### PR TITLE
Feature: auth backend + popup frontend

### LINKED ISSUE #6: Authentication

### PR DESCRIPTION
**Pull request by brrobert17**
closes #6

### CODE REVIEW

**Comment by Roman-Octavian** [comment]
@brrobert17 Need help fixing these conflicts?

**Comment by brrobert17** [comment]
I believe I resolved the conflicts

File: server/routers/userRouter.ts [hunk 1]
```diff
@@ -1,9 +1,12 @@
 import { Router } from 'express';
 import { User } from '../database';
+import { authenticator } from '../utils/auth';
 
 const router = Router();
+// This implementation makes all this router's routes protected, you can also include it with individual endpoints if you like
+router.use(authenticator);
```
Comment (Roman-Octavian) [inline server/routers/userRouter.ts:7]: Can we protect the storage router and the `storage-example.html` page as an example of how to use this?

File: client/index.ts [hunk 2]
```diff
@@ -1,13 +1,17 @@
...
+//   setInterval(() => {
+//     if (h1.className.includes('left-eye')) {
+//       h1.innerHTML = '(⊙_◎)';
+//       h1.className = h1.className.replace('left-eye', 'right-eye');
+//     } else {
+//       h1.innerHTML = '(◎_⊙)';
+//       h1.className = h1.className.replace('right-eye', 'left-eye');
+//     }
+//   }, 500);
+// }
```
Comment (Roman-Octavian) [inline client/index.ts:15]: just remove it :)

File: client/index.html [hunk 3]
```diff
@@ -19,6 +19,8 @@ <h1 id="h1" class="flex text-white xs:text-4xl sm:text-7xl md:text-9xl left-eye"
       >
         Storage Example
       </button>
+      <!-- Button to open the modal -->
```
Comment (Roman-Octavian) [inline client/index.html:22]: Get rid of the H1 above now that we have actual content to show

File: server/routers/authRouter.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    res.json({ message: 'user not found' });
+    return;
+  }
+
+  if (user && (await bcrypt.compare(password, user.password))) {
+    const payLoad = {
+      username: user.username,
+    };
+
+    const accessToken = jwt.sign(payLoad, process.env.JWT_ACCESS_SECRET, {
```
Comment (Roman-Octavian) [inline server/routers/authRouter.ts:25]: Give us an updated `.env` with all these secrets on discord. Also, update the `.template.env` with the changes that you made

File: package.json [hunk 5]
```diff
@@ -22,6 +22,7 @@
   ],
   "license": "MIT",
   "devDependencies": {
+    "axios": "^1.7.7",
```
Comment (Roman-Octavian) [inline package.json:25]: `axios` is not a `devDependency` because it is needed at runtime, not as a development tool.

Also, don't include dependencies that are needed to run the apps inside of the root `package.json`. If you were to want to deploy these projects separately say on Azure or Vercel or whatever, you would want the projects to be self-contained within their directories, because client and server needn't and shouldn't have knowledge of each other. 

If you need to share code specifically across projects, you can use the `packages` directory to create a new module and import that module in both projects separately. But for dependencies, they should never be shared. So install `axios` in both `client` and `server` separately.

File: client/vite-env.d.ts [hunk 6]
```diff
@@ -0,0 +1,7 @@
+// vite-env.d.ts
+/// <reference types="vite/client" />
+
+declare module '*.html?raw' {
+  const content: string;
+  export default content;
+}
```
Comment (Roman-Octavian) [inline client/vite-env.d.ts:7]: Just curious, why did you need to add this?

**Comment by brrobert17** [comment]
to be honest I did't think too much about client side protected routes, but I'll figure something out

File: client/vite-env.d.ts (hunk 6 above)
Comment (brrobert17) [reply client/vite-env.d.ts:7]: it's necessary  in order to import and use raw HTML from client/components/popup/index.HTML

File: client/components/popup/index.ts [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+      loginBtn.innerText = 'Login';
+      toggleForm.innerHTML = 'SIGN UP HERE';
+      nameField.classList.add('hidden');
+      toggleText.innerText = 'DON’T HAVE AN ACCOUNT?';
+    } else {
+      modalTitle.innerText = 'SIGN UP';
+      loginBtn.innerText = 'Sign Up';
+      toggleForm.innerHTML = 'LOG IN HERE';
+      nameField.classList.remove('hidden');
+      toggleText.innerText = 'ALREADY HAVE AN ACCOUNT?';
```
Comment (nikidkny) [inline client/components/popup/index.ts:49]: you should add the option then to change to the log in form

File: client/components/popup/index.ts [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+    let nameInput: HTMLInputElement | undefined;
+    if (!isLogin) {
+      nameInput = document.getElementById('name') as HTMLInputElement;
+    }
+
+    const credentials: { username: string; password: string; name?: string } = {
+      username: usernameInput.value,
+      password: passwordInput.value,
+    };
```
Comment (nikidkny) [inline client/components/popup/index.ts:79]: I think we should add some validation, like it needs to be 8 characters long and has numbers so on

File: client/components/popup/index.ts [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+    if (!isLogin && nameInput) {
+      credentials.name = nameInput.value;
+    }
+
+    try {
+      const endpoint = isLogin ? '/login' : '/signup';
+      const result = await axiosLoginInstance.post(endpoint, credentials);
+
+      alert(result.data.message); // Show success message
+      hideModal(); // Hide the modal after success
```
Comment (nikidkny) [inline client/components/popup/index.ts:90]: I think we should empty the input fields as well after it was successful

**Review by nikidkny — APPROVED** [review]
Smaller screens look a bit funky but otherwise I think looks good
